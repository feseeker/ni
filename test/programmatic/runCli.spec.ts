import path from 'node:path'
import { tmpdir } from 'node:os'
import fs from 'fs-extra'
import { type SpyInstance } from 'vitest'
import { afterAll, beforeAll, describe, expect, test, vi } from 'vitest'
import type { Runner } from '../../src'
import { AGENTS, parseNi, runCli } from '../../src'

let basicLog: SpyInstance, errorLog: SpyInstance, warnLog: SpyInstance, infoLog: SpyInstance

function runCliTest(fixtureName: string, agent: string, runner: Runner, args: string[]) {
  return async () => {
    const cwd = await fs.mkdtemp(path.join(tmpdir(), 'ni-'))
    const fixture = path.join(__dirname, '..', 'fixtures', fixtureName, agent)
    await fs.copy(fixture, cwd)

    await runCli(async (agent, _, ctx) => {
      return runner(agent, args, ctx)
    }, {
      programmatic: true,
      cwd,
    }).catch((e) => {
      if (e.command)
        expect(e.command).toMatchSnapshot()
      else
        expect(e.message).toMatchSnapshot()
    })
  }
}

beforeAll(() => {
  basicLog = vi.spyOn(console, 'log')
  warnLog = vi.spyOn(console, 'warn')
  errorLog = vi.spyOn(console, 'error')
  infoLog = vi.spyOn(console, 'info')

  vi.mock('execa', async (importOriginal) => {
    const mod = await importOriginal() as any
    return {
      ...mod,
      execaCommand: (cmd: string) => {
        // break execution flow for easier snapshotting
        // eslint-disable-next-line no-throw-literal
        throw { command: cmd }
      },
    }
  })
})

afterAll(() => {
  vi.resetAllMocks()
})

const agents = [...Object.keys(AGENTS), 'unknown']
const fixtures = ['lockfile', 'packager']

// matrix testing of: fixtures x agents x commands
fixtures.forEach(fixture => describe(fixture, () => agents.forEach(agent => describe(agent, () => {
/** ni */
  test('ni', runCliTest(fixture, agent, parseNi, []))
  test('ni foo', runCliTest(fixture, agent, parseNi, ['foo']))
  test('ni foo -D', runCliTest(fixture, agent, parseNi, ['foo', '-D']))
  test('ni --frozen', runCliTest(fixture, agent, parseNi, ['--frozen']))
  test('ni -g foo', runCliTest(fixture, agent, parseNi, ['-g', 'foo']))

  test('no logs', () => {
    expect(basicLog).not.toHaveBeenCalled()
    expect(warnLog).not.toHaveBeenCalled()
    expect(errorLog).not.toHaveBeenCalled()
    expect(infoLog).not.toHaveBeenCalled()
  })
}))))
