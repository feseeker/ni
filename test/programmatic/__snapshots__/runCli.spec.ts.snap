// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`lockfile > npm > ni --frozen 1`] = `"npm ci"`;

exports[`lockfile > npm > ni -g foo 1`] = `"npm i -g foo"`;

exports[`lockfile > npm > ni 1`] = `"npm i "`;

exports[`lockfile > npm > ni foo -D 1`] = `"npm i foo -D"`;

exports[`lockfile > npm > ni foo 1`] = `"npm i foo"`;

exports[`lockfile > pnpm > ni --frozen 1`] = `"pnpm i --frozen-lockfile"`;

exports[`lockfile > pnpm > ni -g foo 1`] = `"pnpm add -g foo"`;

exports[`lockfile > pnpm > ni 1`] = `"pnpm i "`;

exports[`lockfile > pnpm > ni foo -D 1`] = `"pnpm add foo -D"`;

exports[`lockfile > pnpm > ni foo 1`] = `"pnpm add foo"`;

exports[`lockfile > pnpm@6 > ni --frozen 1`] = `"pnpm i --frozen-lockfile"`;

exports[`lockfile > pnpm@6 > ni -g foo 1`] = `"pnpm add -g foo"`;

exports[`lockfile > pnpm@6 > ni 1`] = `"pnpm i "`;

exports[`lockfile > pnpm@6 > ni foo -D 1`] = `"pnpm add foo -D"`;

exports[`lockfile > pnpm@6 > ni foo 1`] = `"pnpm add foo"`;

exports[`lockfile > unknown > ni --frozen 1`] = `"npm ci"`;

exports[`lockfile > unknown > ni -g foo 1`] = `"npm i -g foo"`;

exports[`lockfile > unknown > ni 1`] = `"npm i "`;

exports[`lockfile > unknown > ni foo -D 1`] = `"npm i foo -D"`;

exports[`lockfile > unknown > ni foo 1`] = `"npm i foo"`;

exports[`lockfile > yarn > ni --frozen 1`] = `"yarn install --frozen-lockfile"`;

exports[`lockfile > yarn > ni -g foo 1`] = `"yarn global add foo"`;

exports[`lockfile > yarn > ni 1`] = `"yarn install "`;

exports[`lockfile > yarn > ni foo -D 1`] = `"yarn add foo -D"`;

exports[`lockfile > yarn > ni foo 1`] = `"yarn add foo"`;

exports[`lockfile > yarn@berry > ni --frozen 1`] = `"yarn install --frozen-lockfile"`;

exports[`lockfile > yarn@berry > ni -g foo 1`] = `"yarn global add foo"`;

exports[`lockfile > yarn@berry > ni 1`] = `"yarn install "`;

exports[`lockfile > yarn@berry > ni foo -D 1`] = `"yarn add foo -D"`;

exports[`lockfile > yarn@berry > ni foo 1`] = `"yarn add foo"`;

exports[`packager > npm > ni --frozen 1`] = `"npm ci"`;

exports[`packager > npm > ni -g foo 1`] = `"npm i -g foo"`;

exports[`packager > npm > ni 1`] = `"npm i "`;

exports[`packager > npm > ni foo -D 1`] = `"npm i foo -D"`;

exports[`packager > npm > ni foo 1`] = `"npm i foo"`;

exports[`packager > pnpm > ni --frozen 1`] = `"pnpm i --frozen-lockfile"`;

exports[`packager > pnpm > ni -g foo 1`] = `"pnpm add -g foo"`;

exports[`packager > pnpm > ni 1`] = `"pnpm i "`;

exports[`packager > pnpm > ni foo -D 1`] = `"pnpm add foo -D"`;

exports[`packager > pnpm > ni foo 1`] = `"pnpm add foo"`;

exports[`packager > pnpm@6 > ni --frozen 1`] = `"pnpm i --frozen-lockfile"`;

exports[`packager > pnpm@6 > ni -g foo 1`] = `"pnpm add -g foo"`;

exports[`packager > pnpm@6 > ni 1`] = `"pnpm i "`;

exports[`packager > pnpm@6 > ni foo -D 1`] = `"pnpm add foo -D"`;

exports[`packager > pnpm@6 > ni foo 1`] = `"pnpm add foo"`;

exports[`packager > unknown > ni --frozen 1`] = `"npm ci"`;

exports[`packager > unknown > ni -g foo 1`] = `"npm i -g foo"`;

exports[`packager > unknown > ni 1`] = `"npm i "`;

exports[`packager > unknown > ni foo -D 1`] = `"npm i foo -D"`;

exports[`packager > unknown > ni foo 1`] = `"npm i foo"`;

exports[`packager > yarn > ni --frozen 1`] = `"yarn install --frozen-lockfile"`;

exports[`packager > yarn > ni -g foo 1`] = `"yarn global add foo"`;

exports[`packager > yarn > ni 1`] = `"yarn install "`;

exports[`packager > yarn > ni foo -D 1`] = `"yarn add foo -D"`;

exports[`packager > yarn > ni foo 1`] = `"yarn add foo"`;

exports[`packager > yarn@berry > ni --frozen 1`] = `"yarn install --immutable"`;

exports[`packager > yarn@berry > ni -g foo 1`] = `"npm i -g foo"`;

exports[`packager > yarn@berry > ni 1`] = `"yarn install "`;

exports[`packager > yarn@berry > ni foo -D 1`] = `"yarn add foo -D"`;

exports[`packager > yarn@berry > ni foo 1`] = `"yarn add foo"`;
